### PR TITLE
feat: Add advanced Conversation Filters

### LINKED ISSUE #3188: Create Modal Component, payload query builder, with validations.
To create a query as per #3186, we need to create a component in the frontend. 

design-2(3)

**Figma Link:**
https://www.figma.com/file/aZJsiXHfKfJYeFHkS2ph14/Product-Design-Specs?node-id=0%3A1

### PR DESCRIPTION
**Pull request by fayazara**
# Pull Request Template

## Description

Adds a new advanced conversations filter feature which allows users to filter conversations via Status, Assignees, Contacts, Inbox, Teams, Conversation Identifier, Campaigns, Labels, Browser languages, Referrer links. 

Fixes #3188 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Deployed on qa environment to test the feature, specs written for the new api endpoints and payload generators.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/javascript/dashboard/assets/scss/widgets/_modal.scss [hunk 1]
```diff
@@ -89,6 +86,11 @@
     button {
       font-size: $font-size-small;
     }
+
+    &.justify-content-end {
+      justify-content: end;
+      width: 100%;
+    }
```
Comment (pranavrajs) [inline app/javascript/dashboard/assets/scss/widgets/_modal.scss:93]: Please move this to helper file, and use w-full for width:100%

File: app/javascript/dashboard/components/ChatList.vue [hunk 2]
```diff
@@ -295,4 +349,24 @@ export default {
...
+  align-items: center;
+  .btn-filter {
+    margin: 0 var(--space-normal);
+    cursor: pointer;
+    i {
+      font-size: 2rem;
+    }
+  }
+  .btn-clear-filters {
+    color: tomato;
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/ChatList.vue:363]: Please use a color from shared/assets/stylesheets/colors.scss

File: app/javascript/dashboard/components/ChatList.vue [hunk 3]
```diff
@@ -295,4 +349,24 @@ export default {
...
 }
+.filter--actions {
+  display: flex;
+  align-items: center;
+  .btn-filter {
+    margin: 0 var(--space-normal);
+    cursor: pointer;
+    i {
+      font-size: 2rem;
+    }
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/ChatList.vue:360]: Please use a font-size from standard font-sizes.

File: app/javascript/dashboard/components/ChatList.vue [hunk 4]
```diff
@@ -295,4 +349,24 @@ export default {
...
+  }
+  .btn-clear-filters {
+    color: tomato;
+    cursor: pointer;
+  }
+}
+
+.filter__applied {
+  padding: var(--space-slab) 0 !important;
+  border-bottom: 1px solid $color-border;
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/ChatList.vue:370]: Please use the CSS variable. https://www.chatwoot.com/docs/handbook/engineering/frontend-guidelines#use-css-variables-instead-of-scss-variables

File: app/javascript/dashboard/store/modules/conversations/actions.js [hunk 5]
```diff
@@ -52,6 +56,21 @@ const actions = {
     }
   },
 
+  fetchFilteredConversations: async ({ commit, dispatch }, payload) => {
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/conversations/actions.js:59]: Please add a spec for this action.

File: app/javascript/dashboard/assets/scss/widgets/_modal.scss [hunk 6]
```diff
@@ -89,6 +86,11 @@
     button {
       font-size: $font-size-small;
     }
+
+    &.justify-content-end {
```
Comment (nithindavid) [inline app/javascript/dashboard/assets/scss/widgets/_modal.scss:90]: I think the class name is misleading, Can we either rename class by it's nature or split into two classes?

File: app/javascript/dashboard/components/ChatList.vue [hunk 7]
```diff
@@ -1,14 +1,31 @@
...
         {{ pageTitle }}
       </h1>
-      <chat-filter @statusFilterChange="updateStatusType" />
+
+      <div class="filter--actions">
+        <chat-filter
+          v-if="!filtersApplied"
+          @statusFilterChange="updateStatusType"
+        />
+        <button v-else class="btn-clear-filters" @click="resetAndFetchData">
```
Comment (nithindavid) [inline app/javascript/dashboard/components/ChatList.vue:14]: Use `woot-button`

Comment (nithindavid) [reply app/javascript/dashboard/components/ChatList.vue:14]: For all the other occurrences also.

**Comment by fayazara** [comment]
@pranavrajs @nithindavid Worked on the changes.

File: app/javascript/dashboard/components/widgets/conversation/ChatAdvancedFilter.vue [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+      return type.filterOperators;
+    },
+    // eslint-disable-next-line consistent-return
+    getDropdownValues(type) {
+      switch (type) {
+        case 'status':
+          return [
+            {
+              id: 'open',
+              name: 'Open',
```
Comment (muhsin-k) [inline app/javascript/dashboard/components/widgets/conversation/ChatAdvancedFilter.vue:123]: i18n

File: app/javascript/dashboard/i18n/locale/en/advancedFilters.json [hunk 9]
```diff
@@ -0,0 +1,24 @@
+{
+    "FILTER": {
+        "SUBTITLE": "Add filters below and hit 'Submit' to filter conversations.",
+        "ADD_NEW_FILTER": "Add Filter",
+        "FILTER_DELETE_ERROR": "You should have atleast one filter to save",
```
Comment (muhsin-k) [inline app/javascript/dashboard/i18n/locale/en/advancedFilters.json:5]:
```suggestion
        "FILTER_DELETE_ERROR": "You should have at least one filter to save",
```

File: app/javascript/dashboard/components/ChatList.vue [hunk 10]
```diff
@@ -273,6 +340,8 @@ export default {
 
 <style scoped lang="scss">
 @import '~dashboard/assets/scss/woot';
+@import '~widget/assets/scss/variables.scss';
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/ChatList.vue:343]: This is supposed to be used in widget pack only. Please remove this and use CSS variables from `shared/assets/stylesheets/`

File: app/javascript/dashboard/components/widgets/conversation/ConversationAdvancedFilter.vue [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+    },
+    clearPreviousValues(index) {
+      this.appliedFilters[index].values = '';
+    },
+  },
+};
+</script>
+
+<style lang="scss">
+@import '~widget/assets/scss/variables.scss';
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/widgets/conversation/ConversationAdvancedFilter.vue:193]: This is supposed to be used in widget pack only. Please remove this and use CSS variables from shared/assets/stylesheets/

**Review by pranavrajs — CHANGES_REQUESTED** [review]
@fayazara Made some changes in the i18n files, also please see the comment about the CSS variables.

File: app/javascript/dashboard/components/widgets/conversation/ConversationAdvancedFilter.vue (hunk 11 above)
Comment (fayazara) [reply app/javascript/dashboard/components/widgets/conversation/ConversationAdvancedFilter.vue]: Resolved in https://github.com/chatwoot/chatwoot/pull/3239/commits/84c44831e64e65e2efc240708bb12ceea26e22c2

File: app/javascript/dashboard/components/ChatList.vue (hunk 10 above)
Comment (fayazara) [reply app/javascript/dashboard/components/ChatList.vue]: https://github.com/chatwoot/chatwoot/pull/3239/commits/84c44831e64e65e2efc240708bb12ceea26e22c2

**Comment by tejaswinichile** [comment]
Tested the branch on local again with all the new updates. Working fine 👍🏼

**Comment by pranavrajs** [comment]
- [ ] Remove contains, does not contain from Assignee Name, Team Name, Labels, Inbox Name, Campaign Name.
- [ ] Filter on conversation identifier should work on display_id not id
- [ ] Remove selection option from is_present, is_not_present filters.

**Pull request merged**